streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.IntegerSerializer;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.MockTime;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.KafkaStreams.State;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.integration.utils.EmbeddedKafkaCluster;
+import org.apache.kafka.streams.integration.utils.IntegrationTestUtils;
+import org.apache.kafka.streams.kstream.Consumed;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.QueryableStoreTypes;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
+import org.apache.kafka.test.IntegrationTest;
+import org.apache.kafka.test.TestUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(IntegrationTest.class)
+public class OptimizedKTableIntegrationTest {
+    private static final int NUM_BROKERS = 1;
+
+    private static final String INPUT_TOPIC_NAME = "input-topic";
+    private static final String TABLE_NAME = "source-table";
+
+    @Rule
+    public final EmbeddedKafkaCluster cluster = new EmbeddedKafkaCluster(NUM_BROKERS);
+
+    private final Map<KafkaStreams, State> clientStates = new HashMap<>();
+    private final Lock clientStatesLock = new ReentrantLock();
+    private final Condition clientStateUpdate = clientStatesLock.newCondition();
+    private final MockTime mockTime = cluster.time;
+
+    @Before
+    public void before() throws InterruptedException {
+        cluster.createTopic(INPUT_TOPIC_NAME, 2, 1);
+    }
+
+    @After
+    public void after() {
+        for (final KafkaStreams client : clientStates.keySet()) {
+            client.close();
+        }
+    }
+
+    @Test
+    public void standbyShouldNotPerformRestoreAtStartup() throws InterruptedException, ExecutionException {
+        final int numMessages = 10;
+        final int key = 1;
+        final Semaphore semaphore = new Semaphore(0);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder
+            .table(INPUT_TOPIC_NAME, Consumed.with(Serdes.Integer(), Serdes.Integer()),
+                Materialized.<Integer, Integer, KeyValueStore<Bytes, byte[]>>as(TABLE_NAME)
+                    .withCachingDisabled())
+            .toStream()
+            .peek((k, v) -> semaphore.release());
+
+        final KafkaStreams client1 = createClient(builder, streamsConfiguration());
```

Comment:
nit: `client1` -> `kafkaStreams1`
nit `createClient` -> `createKafkaStreams`

(similar for all other variables eg `clientStates` etc)

It's technically a client, but also consumer and producer are and we also use `consumer`/`producer` as variable name there -- makes it easier to read -- otherwise it unclear what client it is.

Reply:
Will do. Thanks!

Reply:
Any objection to calling it an instance?

Reply:
While I think it would be fine, it would be "new" -- if we think `instance` is better, we might want to migrate all other code lazily to use `instance`, too. It's always best to have the same naming conventions throughout the whole codebase IMHO.

Reply:
+1

I liked the name `instance` because it matched the documentation and was unambiguous. On the other hand, we're actually creating an object of type `KafkaStreams`, so that is a pretty strong argument to use the convention you originally suggested.

I'll switch instanceXXX to kafkaStreamsXXX and submit a follow up patch.